### PR TITLE
Make dynamo enable_nnmodule_hooks=False by default

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98253
* #97184
* __->__ #98223

Previously, hook support was enabled globally and not configurable.

Since guard overhead due to hooks is significant (and some hook related
features are missing or WIP), and hooks used less often than not used,
it makes more sense to disable by default at least for now.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -122,6 +122,13 @@ def __init__(self, fn, is_constant=False, **kwargs):
         ), f"expected FunctionType found {typestr(fn)} {fn}"
         # unpack @torch._dynamo.optimize()(fn) wrapped function
         fn = inspect.getattr_static(fn, "_torchdynamo_inline", fn)
+        if istype(fn, types.MethodType):
+            # when we record the original fn, it may be a bound method such as .forward
+            # however, NNModuleVariable.call_function special-cases .forward to be called
```
Comment (wconstab) [inline torch/_dynamo/variables/functions.py:127]: i'm working on pushing out another PR on the top of the stack to reverse this and try using Method in nnmodule, naive approach bombed out somewhere..

File: test/dynamo/test_modules.py [hunk 2]
```diff
@@ -1341,8 +1344,6 @@ def guard_fail_fn(failure):
             backend=cc,
         )(outer_func)
 
-        m = TestModule()
```
Comment (wconstab) [inline test/dynamo/test_modules.py:1344]: intentional deletion, this code looks to have been an oversight/copypaste-error when i put the test up in the first place.  constructing the new module and registering the hook (when the module/hook were created above) is pointless and not part of the test case.

**Review by jansel — APPROVED** [review]

**Comment by wconstab** [comment]
@jansel I just realized this has a downside of removing CI coverage for the enabled path.  Is it worth adding a separate CI step to repeat the dynamo tests (+benchmarks) with hooks enabled?

**Comment by jansel** [comment]
I don't think we need to repeat all tests (most are just functions not nn.Modules), though adding a few targeted tests makes sense.

**Comment by wconstab** [comment]
> I don't think we need to repeat all tests (most are just functions not nn.Modules), though adding a few targeted tests makes sense.

I don't think we need to repeat all the opinfo tests or some of the other auxiliary stuff, but during hooks enablement i found failures popped out of the woodwork across various benchmark models and different suites underneath test/dynamo besides test_nn_modules.  It's actually a pretty sensitive codepath

**Comment by jansel** [comment]
You could do something similar to our dynamic tests which repeat entire test suites.

**Comment by wconstab** [comment]
yea, i am happy with copying test_dynamic_shapes for the unit test part.

The benchmark part is more problematic though.
* its about 1-2hrs per suite to run accuracy bench, and we already forked it to run 4 suites x (static/dynamic)
* we should really keep perf-ci and accuracy ci totally aligned (e.g. don't just change accuracy CI to run with hooks and perf to skip them)
* its not great to offer the 'enable hooks' feature if we're not offering poor CI coverage

If we think hooks is really an uncommon use case, we could forego the benchmark coverage and just try to keep adding unit tests when we find gaps.  I don't really want to add another fork to all the benchmark accuracy CI though, but open to others thoughts.  cc @ezyang @albanD @desertfire

**Comment by jansel** [comment]
AFAIK nothing in our accuracy bench suite uses hooks, so I don't see how running it would be helpful.  There are also dozens of other flags we don't run our benchmark suites with.

**Comment by wconstab** [comment]
Opacus actually does run with hooks and is currently running in a 'broken' state bc it currently skips hooks only on 'allowed-in-graph' modules such as .conv.  I haven't checked if other benchmarks have hooks, but i agree it's not the common path.

However, the changes made to make hooks work affect the codepath of all models- so without running all the accuracy bench on the hooks path, it's easy to accidentally break the hooks path without realizing it.  (I broke quite a few benchmark models on my PRs thankfully caught by CI over the various iterations of enabling hooks in the first place).

**Comment by jansel** [comment]
Should we just run `Opacus` then?

**Comment by wconstab** [comment]
I am thinking of abandoning this approach and instead, default to tracing __call__  impl but also 'skip hook guards', and have dynamo patch hook-dicts and hook setters on modules to either error or automatically enable guards and recompile.

**Comment by wconstab** [comment]
abandoned in favor of https://github.com/pytorch/pytorch/pull/98371
